src/core/algorithms/dc/model/column_operand.h

```diff
@@ -3,79 +3,136 @@
+    std::optional<dc::Tuple> tuple_;
+    model::Type const* type_;
+    std::byte const* val_;
 
 public:
-    ColumnOperand(Column const* column, bool is_first_tuple) noexcept
-        : column_(column), is_first_tuple_(is_first_tuple) {}
-
-    ColumnOperand() noexcept = default;
-
-    // For conversion from "t.ColumnPosition" or "t.ColumnName"
-    ColumnOperand(std::string operand, RelationalSchema const& schema) {
-        if (operand.front() != 't' and operand.front() != 's')
-            throw std::logic_error("Unknown tuple name");
-
-        is_first_tuple_ = operand.front() == 't';
-        std::string name(operand.begin() + 2, operand.end());
-        std::vector<std::unique_ptr<Column>> const& cols = schema.GetColumns();
-        if (!cols.front()->GetName().empty()) {  // Has header
-            for (std::unique_ptr<Column> const& col : cols) {
-                if (name == col->GetName()) {
-                    column_ = col.get();
-                    return;
-                }
-            }
+    ColumnOperand() noexcept : column_(nullptr), val_(nullptr) {};
+
+    ColumnOperand(Column const* column, dc::Tuple tuple, model::Type const* type)
+        : column_(column), tuple_(tuple), type_(type), val_(nullptr) {}
+
+    ColumnOperand(std::string const& str_val, model::Type const* type)
+        : column_(nullptr), type_(type) {
+        std::byte* val = type->Allocate();
+        type->ValueFromStr(val, str_val);
+        val_ = val;
+    }
+
+    ColumnOperand(ColumnOperand const& rhs) : type_(rhs.type_) {
+        if (rhs.IsVariable()) {
+            tuple_ = rhs.tuple_;
+            column_ = rhs.column_;
+            val_ = nullptr;
+        } else {
+            val_ = rhs.type_->Clone(rhs.val_);
         }
+    }
 
-        std::string str_ind(operand.begin() + 2, operand.end());
-        model::ColumnIndex ind = std::stoi(str_ind);
-        column_ = cols[ind].get();
+    ColumnOperand(ColumnOperand&& rhs) : ColumnOperand() {
+        Swap(rhs);
     }
 
-    bool operator==(ColumnOperand const& rhs) const noexcept {
-        return column_ == rhs.column_ && is_first_tuple_ == rhs.is_first_tuple_;
+    ColumnOperand& operator=(ColumnOperand rhs) {
+        Swap(rhs);
+        return *this;
+    }
+
+    void Swap(ColumnOperand& rhs) {
+        std::swap(type_, rhs.type_);
+        std::swap(val_, rhs.val_);
+        std::swap(column_, rhs.column_);
+        std::swap(tuple_, rhs.tuple_);
+    }
+
+    bool operator==(ColumnOperand const& rhs) const {
+        if (IsConstant() != rhs.IsConstant()) return false;
+
+        if (IsConstant()) {
+            assert(type_ == rhs.type_);
+            return type_->Compare(GetVal(), rhs.GetVal()) == model::CompareResult::kEqual;
+        }
+
+        return column_ == rhs.column_ && tuple_ == rhs.tuple_;
     }
 
-    bool operator!=(ColumnOperand const& rhs) const noexcept {
+    bool operator!=(ColumnOperand const& rhs) const {
         return !(*this == rhs);
     }
```

Comment:
I think C++20 can do this automatically with `bool operator!=(ColumnOperand const& rhs) = default` assuming `operator==` is provided